packages/compass-data-modeling/src/services/export-diagram.tsx

```diff
@@ -67,33 +35,50 @@ function moveSvgDefsToViewportElement(
   markerDef.remove();
 }
 
-export async function exportToPng(fileName: string, diagram: DiagramInstance) {
-  const container = document.createElement('div');
-  container.setAttribute('data-testid', 'export-diagram-container');
-  // Push it out of the viewport
-  container.style.position = 'fixed';
-  container.style.top = '100vh';
-  container.style.left = '100vw';
-  document.body.appendChild(container);
-
-  const dataUri = await getExportPngDataUri(container, diagram);
-  downloadFile(dataUri, fileName, () => {
-    container.remove();
-  });
+export async function exportToPng(
+  fileName: string,
+  diagram: DiagramInstance,
+  signal?: AbortSignal
+) {
+  const dataUri = await raceWithAbort(
+    getExportPngDataUri(diagram),
+    signal ?? new AbortController().signal
+  );
+  downloadFile(dataUri, fileName);
 }
 
-export function getExportPngDataUri(
-  container: HTMLDivElement,
-  diagram: DiagramInstance
-): Promise<string> {
-  return new Promise<string>((resolve, reject) => {
+export function getExportPngDataUri(diagram: DiagramInstance): Promise<string> {
+  return new Promise<string>((resolve, _reject) => {
```

Comment:
Is there a reason why we're going to the Promise resolve/reject here? I'm thinking it could lead to an uncaught exception. Could we instead have the cleanup in a finally or catch block?

We could do await something like the rafraf as well and await the `toPng`.

Reply:
What would you recommend in this case? I only want to resolve once dom has been converted to a data uri (and as such show loading state to the user). For clean up, its already in place where we export.

Reply:
I would recommend async/await here with try/catch. I find it makes the code easier to read through and the error handling less prone to be uncaught.